src/vsock/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+        }
+
+        let mut thread = self.threads[thread_id].lock().unwrap();
+        let evt_idx = thread.event_idx;
+
+        match device_event {
+            RX_QUEUE_EVENT => {
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            TX_QUEUE_EVENT => {
+                thread.process_tx(vring_tx_lock, evt_idx)?;
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            EVT_QUEUE_EVENT => {}
+            BACKEND_EVENT => {
+                thread.process_backend_evt(evset);
+                thread.process_tx(vring_tx_lock, evt_idx)?;
+                if thread.thread_backend.pending_rx() {
+                    thread.process_rx(vring_rx_lock, evt_idx)?;
+                }
+            }
+            _ => {
+                return Err(Error::HandleUnknownEvent.into());
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn get_config(&self, _offset: u32, _size: u32) -> Vec<u8> {
+        let buf = unsafe {
+            slice::from_raw_parts(
+                &self.guest_cid as *const __u64 as *const _,
+                mem::size_of::<__u64>(),
+            )
+        };
+        buf.to_vec()
+    }
+
+    fn queues_per_thread(&self) -> Vec<u64> {
+        self.queues_per_thread.clone()
+    }
+}
+
+/// This is the public API through which an external program starts the
+/// vhost-user-vsock backend server.
+pub fn start_backend_server(vsock_config: VsockConfig) {
+    let vsock_backend = Arc::new(RwLock::new(
+        VhostUserVsockBackend::new(vsock_config.guest_cid, vsock_config.uds_path).unwrap(),
+    ));
+
+    let listener = Listener::new(vsock_config.socket, true).unwrap();
+
+    let name = "vhost-user-vsock";
+    let mut vsock_daemon = VhostUserDaemon::new(name.to_string(), vsock_backend.clone()).unwrap();
+
+    let mut vring_workers = vsock_daemon.get_vring_workers();
+
+    if vring_workers.len() != vsock_backend.read().unwrap().threads.len() {
+        error!("Number of vring workers must be identical to number of backend threads");
+    }
+
+    for thread in vsock_backend.read().unwrap().threads.iter() {
+        thread
+            .lock()
+            .unwrap()
+            .set_vring_worker(Some(vring_workers.remove(0)));
+    }
+    if let Err(e) = vsock_daemon.start(listener) {
+        dbg!("Failed to start vsock daemon: {:?}", e);
+        process::exit(1);
+    }
+
+    if let Err(e) = vsock_daemon.wait() {
+        dbg!("Error from daemon main thread: {:?}", e);
+        process::exit(1);
+    }
+
+    debug!("Vsock daemon is finished");
+
+    for thread in vsock_backend.read().unwrap().threads.iter() {
+        if let Err(e) = thread.lock().unwrap().kill_evt.write(1) {
+            error!("Error shutting down worker thread: {:?}", e);
+        }
+    }
```

Comment:
Try this, start this dameon and then start your qemu instance. Once you kill the qemu instance, you won't be able to use the daemon anymore without restarting it after killing it first. I think this can be improved a bit.

A daemon, in my opinion, should just run for ever in background, without the need of ever restarting it ever.

If you look at how I did this in case of I2C, you will get an idea.